### PR TITLE
[skip-ci][ntuple] specifications.md: document `RNTupleCardinality`

### PR DESCRIPTION
**Pull request by jalopezg-git**
See also the former PR that introduces this projected field: #12008.

## Checklist:
- [X] updated the docs (if necessary)

### CODE REVIEW

File: tree/ntuple/v7/doc/specifications.md [hunk 1]
```diff
@@ -677,6 +677,13 @@ The on-disk representation is similar to a `std::vector<T>` where `T` is the val
   - Child field of type `T`, which must by a type with RNTuple I/O support.
     The name of the child field is `_0`.
 
+### ROOT::Experimental::RNTupleCardinality
+
+A field whose type is `ROOT::Experimental::RNTupleCardinality` should always appear as a "projected" field whose single alias column is the index column of a collection (see Section "Alias columns").
```
Comment (jblomer) [inline tree/ntuple/v7/doc/specifications.md:682]: I wouldn't restrict ourselves too much. This field can may appear in more contexts and at can be, at some point, a writable field, too. What counts for the specification is that it is represented by an index column, which it presents/interprets as offset lengths (and vice versa: lengths to offsets, should it come to writing).

File: tree/ntuple/v7/doc/specifications.md [hunk 2]
```diff
@@ -677,6 +677,13 @@ The on-disk representation is similar to a `std::vector<T>` where `T` is the val
...
+
+This read-only projected field offers a run-time view over the index column in which each value represents the cardinality of the pointed-to collection.
+The cardinality for the $i$-th element is computed by subtracting the $(i-1)$-th value from the $i$-th value in the index column, e.g. given the index column values `[1, 1, 3]`, the values computed at run-time by the `RNTupleCardinality` field shall be `[1, 0, 2]`.
```
Comment (jblomer) [inline tree/ntuple/v7/doc/specifications.md:685]: I think we need to explain this a bit more: we need to discuss the case where i falls on the start of a cluster (which includes the special case `i == 0`).

**Review by jblomer — CHANGES_REQUESTED** [review]
Many thanks! See comments inline.

File: tree/ntuple/v7/doc/specifications.md (hunk 1 above)
Comment (Moelf) [reply tree/ntuple/v7/doc/specifications.md:682]: https://root.cern.ch/doc/master/RField_8hxx_source.html#l00904

says it's only for reading

**Comment by Moelf** [comment]
tangentially, what's the fundamental reason we need a special interpretation of the same `offset` array instead of just having a new column for tracking the index? can we really not spare one more `Uint16` or even `UInt8` to track it?

(Notice, by using `Cardinality`, the total counter is restricted to `typemax(UInt32) = 4294967295` anyway)

**Comment by jblomer** [comment]
> tangentially, what's the fundamental reason we need a special interpretation of the same `offset` array instead of just having a new column for tracking the counter? can we really not spare one more `Uint16` or even `UInt8` per event to track it?

The reason is indeed space savings. It's something like 1-4 byte per collection (uncompressed), so e.g. with nanoAODs with its ~25 collections, my assumption is that not duplicating the counts allows for one or a few additional real data columns.

By the way, you don't necessarily need to look at the C++ type to distinguish between offsets and vector lengths. You can reason that a _leaf_ field (`role == 0`) backed by an `[Split]Index[32|64]` column should be interpreted as vector lengths and for a _collection_ field (`role == 1`) it's offsets.
 
> (Notice, by using `Cardinality`, the total counter is restricted to `typemax(UInt32) = 4294967295` -> a restriction on average counter * number of events in cluster)

Not sure I see that this is a limitation of the `RNTupleCardinality` type. It is currently 32bit limited due to the use of `Index32` columns, but with `[Split]Index64` columns vector lengths can be 64bit, too.

**Comment by Moelf** [comment]
I thought about this more, which seems to be an outlier of the "leaf field" concept. For almost all leaf fields, you have a column which stores one of the primitive data types, and when you index, you get one element from that column.

maybe the correct way to think about it is it's an additional type of field (like vector, union, struct), except it's marked by combination or `role` and `type` instead of the usual "role is enough".

**Comment by jblomer** [comment]
> I thought about this more, which seems to be an outlier of the "leaf field" concept. For almost all leaf fields, you have a column which stores one of the primitive data types, and when you index, you get one element from that column.

Note that in general, a field (including leaf field) can have [0..$k$] columns. The first column is the "principal column" that is used to identify one element of that field. For instance, the `std::string` leaf field has an index and a char column.

**Comment by Moelf** [comment]
Indeed! a leaf field with `std::string` is special because it's the only thing that "splits" in column records (i.e. the string is basically a jagged vector of `char`, `role=1` jagged vector splits in field records).

Another outlier we will run into someday is `_collection0` with only one child-field, this would look identical to a `vector` (because role=1).

Overall, I think there's something powerful about dealing with type schemas without dealing with value information (columns). And I think RNTuple's type system has the potential and capacity (add more numbers to role) to make it happen.

**Comment by jalopezg-git** [comment]
I have reworded the definition for this field, let me know what you think about it.

**Review by jblomer — APPROVED** [review]

**Pull request merged**